analyzers/src/SonarAnalyzer.Common/Rules/RouteTemplateShouldNotStartWithSlashBase.cs

```diff
@@ -0,0 +1,100 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Concurrent;
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class RouteTemplateShouldNotStartWithSlashBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+{
+    private const string DiagnosticId = "S6931";
+    private const string MessageOnlyActions = "Change the paths of the actions of this controller to be relative and adapt the controller route accordingly.";
+    private const string MessageActionsAndController = "Change the paths of the actions of this controller to be relative and add a controller route with the common prefix.";
+
+    protected abstract TSyntaxKind MethodSyntaxKind { get; }
+    protected abstract Location ControllerLocation(INamedTypeSymbol symbol);
+
+    protected override string MessageFormat => "{0}";
+    protected RouteTemplateShouldNotStartWithSlashBase() : base(DiagnosticId) { }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(compilationStartContext =>
+        {
+            if (!ShouldRegisterAction(compilationStartContext.Compilation))
+            {
+                return;
+            }
+
+            compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+            {
+                var symbol = (INamedTypeSymbol)symbolStartContext.Symbol;
+                if (symbol.IsControllerType())
+                {
+                    var controllerActionInfo = new ConcurrentBag<ControllerActionInfo>();
+                    symbolStartContext.RegisterSyntaxNodeAction(nodeContext =>
+                    {
+                        var methodSymbol = nodeContext.SemanticModel.GetDeclaredSymbol(nodeContext.Node) as IMethodSymbol;
```

Comment:
What do we do with controller methods that are in generated code (e.g. Controllers that are partial classes)? We can
* Completly ignore them
* Include them in the analysis but do not report (that is what we do now)
* Include and report on them

Reply:
I'd include them and not report to them if they are auto-generated.
I'd include and report if they are not auto-generated.

PS: There's another task to deal with partial classes generally.

@antonioaversa  wdyt?

Reply:
Ideally and intuitively, we should neither include nor report on them.
It can come at a later stage, when partial classes support is introduced and tested.
@martin-strecker-sonarsource Do you see any advantage in including them in the analysis, yet not reporting, compared to completely ignoring them? Isn't including but not reporting just a waste of CPU cycles?

Reply:
@antonioaversa One scenario I had in mind is having a partial class that is half autogenerated and half user-defined.
The user-defined one has all actions starting with `/`, but the autogenerated is not.
This should not raise an issue according to the rule.
So I guess we should analyze autogenerated classes too - unless we want to check first if they are autogenerated partially and ignore them otherwise.